File: human-protocol-btc/kotlin/protocol/src/main/kotlin/org/hpb/protocol/Offer.kt
package org.hpb.protocol

import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.hpb.engine.hex
import org.hpb.engine.sha256
import org.hpb.protocol.Pj.a
import org.hpb.protocol.Pj.b
import org.hpb.protocol.Pj.d
import org.hpb.protocol.Pj.i
import org.hpb.protocol.Pj.l
import org.hpb.protocol.Pj.o
import org.hpb.protocol.Pj.s
import org.hpb.engine.nostr.Events
import org.hpb.engine.nostr.NostrEvent

enum class ValidationType { GROUNDTRUTH, AGREEMENT }

/**
 * Manifest-committed mechanical validation policy. Both strategies are pure
 * functions over the revealed submission set — any observer can recompute
 * validation AND payout correctness.
 */
data class ValidationPolicy(
    val type: ValidationType,
    /** GROUNDTRUTH: sha256("<task_key>:<normalized answer>") allowed set. */
    val groundtruthHashes: Set<String> = emptySet(),
    /** AGREEMENT: assignments per task and the majority fraction required. */
    val assignmentsPerTask: Int = 1,
    val agreementThreshold: Double = 0.5,
)

data class KycPolicy(val required: Boolean, val attesters: List<String> = emptyList())

/** A task list carried INLINE in the offer (Nostr-first artifacts). */
data class Task(val key: String, val question: String)

data class JobOffer(
    val escrowId: String,
    val escrowAddress: String,
    val jobType: String,
    val rewardPerTaskSats: Long,
    val tasks: List<Task>,
    val validation: ValidationPolicy,
    val kyc: KycPolicy,
    val expiresAt: Long,
    val status: String = "open",
) {
    /** The manifest string whose sha256 is committed on-chain at setup. */
    fun manifestJson(): String = Pj.obj(
        "v" to Pj.num(ProtocolKinds.VERSION),
        "job_type" to Pj.str(jobType),
        "reward_per_task_sats" to Pj.num(rewardPerTaskSats),
        "tasks" to Pj.arr(
            tasks.map { Pj.obj("key" to Pj.str(it.key), "question" to Pj.str(it.question)) },
        ),
        "validation" to validationJson(),
    ).toString()

    fun manifestHash(): ByteArray = sha256(manifestJson().toByteArray())

    private fun validationJson() = Pj.obj(
        "type" to Pj.str(validation.type.name.lowercase()),
        "groundtruth_hashes" to Pj.arr(validation.groundtruthHashes.sorted().map(Pj::str)),
        "assignments_per_task" to Pj.num(validation.assignmentsPerTask),
        "agreement_threshold" to Pj.num(validation.agreementThreshold),
    )
}

object Offers {
    fun toEvent(privkey: ByteArray, offer: JobOffer, createdAt: Long): NostrEvent =
        Events.sign(
            privkey, ProtocolKinds.JOB_OFFER,
            tags = listOf(
                listOf("d", offer.escrowId),
                listOf("x", offer.escrowId),
                listOf("t", offer.jobType),
            ),
            content = Pj.obj(
                "v" to Pj.num(ProtocolKinds.VERSION),
                "escrow_address" to Pj.str(offer.escrowAddress),
                "reward_per_task_sats" to Pj.num(offer.rewardPerTaskSats),
                "manifest" to Pj.str(offer.manifestJson()),
                "kyc" to Pj.obj(
                    "required" to Pj.bool(offer.kyc.required),
                    "attesters" to Pj.arr(offer.kyc.attesters.map(Pj::str)),
                ),
                "expires_at" to Pj.num(offer.expiresAt),
                "status" to Pj.str(offer.status),
            ).toString(),
            createdAt = createdAt,
        )

    fun fromEvent(event: NostrEvent): JobOffer {
        require(event.kind == ProtocolKinds.JOB_OFFER) { "not a job offer" }
        val content = Pj.parse(event.content)
        val manifest = Pj.parse(content.s("manifest"))
        val validation = manifest.o("validation")
        val kyc = content.o("kyc")
        return JobOffer(
            escrowId = requireNotNull(event.tagValue("d")) { "offer missing d tag" },
            escrowAddress = content.s("escrow_address"),
            jobType = manifest.s("job_type"),
            rewardPerTaskSats = manifest.l("reward_per_task_sats"),
            tasks = manifest.a("tasks").map {
                Task(it.jsonObject.s("key"), it.jsonObject.s("question"))
            },
            validation = ValidationPolicy(
                type = ValidationType.valueOf(validation.s("type").uppercase()),
                groundtruthHashes = validation.a("groundtruth_hashes")
                    .map { it.jsonPrimitive.content }.toSet(),
                assignmentsPerTask = validation.i("assignments_per_task"),
                agreementThreshold = validation.d("agreement_threshold"),
            ),
            kyc = KycPolicy(
                required = kyc.b("required"),
                attesters = kyc.a("attesters").map { it.jsonPrimitive.content },
            ),
            expiresAt = content.l("expires_at"),
            status = content.s("status"),
        )
    }

    /** The committed manifest hash hex (what SETUP carries on-chain). */
    fun manifestHashHex(offer: JobOffer): String = offer.manifestHash().hex()
}

File: ios/HpbCore/Sources/HpbCore/Offer.swift
import Foundation

public enum ValidationType: String {
    case groundtruth
    case agreement
}

/// Manifest-committed mechanical validation policy — pure functions over
/// the revealed submission set.
public struct ValidationPolicy {
    public let type: ValidationType
    public let groundtruthHashes: Set<String>
    public let assignmentsPerTask: Int
    public let agreementThreshold: Double

    public init(
        type: ValidationType,
        groundtruthHashes: Set<String> = [],
        assignmentsPerTask: Int = 1,
        agreementThreshold: Double = 0.5
    ) {
        self.type = type
        self.groundtruthHashes = groundtruthHashes
        self.assignmentsPerTask = assignmentsPerTask
        self.agreementThreshold = agreementThreshold
    }
}

public struct KycPolicy {
    public let required: Bool
    public let attesters: [String]

    public init(required: Bool, attesters: [String] = []) {
        self.required = required
        self.attesters = attesters
    }
}

/// A task list carried INLINE in the offer (Nostr-first artifacts).
public struct TaskItem: Equatable {
    public let key: String
    public let question: String

    public init(key: String, question: String) {
        self.key = key
        self.question = question
    }
}

public struct JobOffer {
    public let escrowId: String
    public let escrowAddress: String
    public let jobType: String
    public let rewardPerTaskSats: Int64
    public let tasks: [TaskItem]
    public let validation: ValidationPolicy
    public let kyc: KycPolicy
    public let expiresAt: Int64
    public let status: String

    public init(
        escrowId: String, escrowAddress: String, jobType: String,
        rewardPerTaskSats: Int64, tasks: [TaskItem], validation: ValidationPolicy,
        kyc: KycPolicy, expiresAt: Int64, status: String = "open"
    ) {
        self.escrowId = escrowId
        self.escrowAddress = escrowAddress
        self.jobType = jobType
        self.rewardPerTaskSats = rewardPerTaskSats
        self.tasks = tasks
        self.validation = validation
        self.kyc = kyc
        self.expiresAt = expiresAt
        self.status = status
    }

    /// The manifest string whose sha256 is committed on-chain at setup.
    public func manifestJson() -> String {
        Cj.write(Cj.obj([
            ("v", Cj.num(ProtocolKinds.version)),
            ("job_type", .str(jobType)),
            ("reward_per_task_sats", Cj.num(rewardPerTaskSats)),
            ("tasks", .arr(tasks.map {
                Cj.obj([("key", .str($0.key)), ("question", .str($0.question))])
            })),
            ("validation", validationJson()),
        ]))
    }

    public func manifestHash() -> [UInt8] {
        sha256(manifestJson())
    }

    private func validationJson() -> J {
        Cj.obj([
            ("type", .str(validation.type.rawValue)),
            ("groundtruth_hashes", .arr(
                validation.groundtruthHashes.sorted(by: utf16Less).map(J.str)
            )),
            ("assignments_per_task", Cj.num(validation.assignmentsPerTask)),
            ("agreement_threshold", Cj.num(validation.agreementThreshold)),
        ])
    }
}

public enum Offers {
    public static func toEvent(_ privkey: [UInt8], _ offer: JobOffer, createdAt: Int64) throws -> NostrEvent {
        try Events.sign(
            privkey, kind: ProtocolKinds.jobOffer,
            tags: [
                ["d", offer.escrowId],
                ["x", offer.escrowId],
                ["t", offer.jobType],
            ],
            content: Cj.write(Cj.obj([
                ("v", Cj.num(ProtocolKinds.version)),
                ("escrow_address", .str(offer.escrowAddress)),
                ("reward_per_task_sats", Cj.num(offer.rewardPerTaskSats)),
                ("manifest", .str(offer.manifestJson())),
                ("kyc", Cj.obj([
                    ("required", .bool(offer.kyc.required)),
                    ("attesters", .arr(offer.kyc.attesters.map(J.str))),
                ])),
                ("expires_at", Cj.num(offer.expiresAt)),
                ("status", .str(offer.status)),
            ])),
            createdAt: createdAt
        )
    }

    public static func fromEvent(_ event: NostrEvent) throws -> JobOffer {
        guard event.kind == ProtocolKinds.jobOffer else { throw HpbError.invalid("not a job offer") }
        let content = try Cj.parse(event.content)
        let manifest = try Cj.parse(content.s("manifest"))
        let validation = try manifest.o("validation")
        let kyc = try content.o("kyc")
        guard let escrowId = event.tagValue("d") else { throw HpbError.invalid("offer missing d tag") }
        guard let type = ValidationType(rawValue: try validation.s("type").lowercased()) else {
            throw HpbError.invalid("unknown validation type")
        }
        return JobOffer(
            escrowId: escrowId,
            escrowAddress: try content.s("escrow_address"),
            jobType: try manifest.s("job_type"),
            rewardPerTaskSats: try manifest.l("reward_per_task_sats"),
            tasks: try manifest.a("tasks").map {
                TaskItem(key: try $0.s("key"), question: try $0.s("question"))
            },
            validation: ValidationPolicy(
                type: type,
                groundtruthHashes: Set(try validation.a("groundtruth_hashes").compactMap(\.stringValue)),
                assignmentsPerTask: try validation.i("assignments_per_task"),
                agreementThreshold: try validation.d("agreement_threshold")
            ),
            kyc: KycPolicy(
                required: try kyc.b("required"),
                attesters: try kyc.a("attesters").compactMap(\.stringValue)
            ),
            expiresAt: try content.l("expires_at"),
            status: try content.s("status")
        )
    }
}
